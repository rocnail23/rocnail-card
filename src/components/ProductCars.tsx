import { useProduct } from "../hooks/useProduct"
import React,{ createContext, ReactElement,CSSProperties } from 'react';
import styles from "../styles/styles.module.css"
import { ContextValues, InitialValues, OnChangeArgs, Product, ProductCardHandlers } from '../interfaces/interfaces';




export const CarsContex = createContext({} as ContextValues)
const {Provider} = CarsContex

export interface ProductProps  {
  product : Product,
  /* children: ReactElement | ReactElement[], */
  children: (event: ProductCardHandlers) => ReactElement
  className?: string,
  style?: CSSProperties,
  onChange?: (args: OnChangeArgs) => void,
  value?: number, 
  initialValues ?: InitialValues
}

export const ProductCars = ({product,children,className,style,onChange,value,initialValues}: ProductProps) => {

  const {increaseBy,state,reset} = useProduct({onChange,product,value,initialValues})
  
  


  return (
    <Provider value={{
        product,
        increaseBy,
        state,
        maxCount: initialValues?.maxCount
    }}>
    <div className={`${styles.productCard} ${className}`}
    style={style}>

      {children({
        count:state,increaseBy,
        product,
        maxCount:initialValues?.maxCount,
        isMaxCountReached: state == initialValues?.maxCount,
        reset

      })}
    </div>
    </Provider>
  )
}


