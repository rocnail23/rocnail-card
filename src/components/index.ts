import { ProductImage } from ".";
import { ProductButtons } from ".";
import { ProductTitle } from ".";
import { ProductCarsHoc } from ".";
import { Hoc1 as PropsHoc } from "../interfaces/interfaces";


export { ProductButtons  } from "./ProductButtons";
export { ProductCars as ProductCarsHoc } from "./ProductCars";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCars:PropsHoc = Object.assign(ProductCarsHoc,{
    title: ProductTitle,
    image: ProductImage,
    buttons: ProductButtons
})

export default ProductCars