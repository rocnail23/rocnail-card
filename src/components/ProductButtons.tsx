import React,{ useCallback, useContext } from "react"
import styles from "../styles/styles.module.css"
import { CarsContex } from "./ProductCars"

interface  Props {
    className?: string
}

export const ProductButtons = ({className}:Props) => {

    const {increaseBy,state,maxCount} = useContext(CarsContex)
    console.log(maxCount)
    const reachMaxCount = useCallback(() => state == maxCount ,[state,maxCount])
    console.log(reachMaxCount())
    return (
        <div className={`${styles.buttonsContainer} ${className}`}>

            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>

            <div className={styles.countLabel}>{state}</div>

            <button className={`${styles.buttonAdd} ${reachMaxCount() && styles.disabled}`} onClick={() => increaseBy(1)}>+</button>

       </div>
    )
}
