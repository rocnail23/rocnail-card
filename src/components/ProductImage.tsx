import React,{ useContext, CSSProperties } from 'react';
import noImage from "../assets/no-image.jpg"
import styles from "../styles/styles.module.css"
import { CarsContex } from "./ProductCars"

export interface Props {
    image?: string,
    className?: string,
    style?: CSSProperties
}

export const ProductImage = ({image,className,style}:Props) => {

    const {product} = useContext(CarsContex)
    let ImageShow: string =   image ? image : product.image ? product.image : noImage

    return (
        <img className={`${styles.productImg} ${className}`} src={ImageShow} alt="coffe mug" 
        style={style}/>
    )
}