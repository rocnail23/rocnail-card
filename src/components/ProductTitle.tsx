import React,{ useContext, CSSProperties } from 'react';
import styles from "../styles/styles.module.css"
import { CarsContex } from "./ProductCars"

export interface Props {title?:string,className?:string,style?: CSSProperties} 

export const ProductTitle = ({title,className,style}:Props) => {

    const {product} = useContext(CarsContex)

    const titleShow: string = title ? title : product.title ? product.title : ""

    return (
        <span className={`${styles.productDescription} ${className}`}
        style={style}>{titleShow}</span>
    )
}