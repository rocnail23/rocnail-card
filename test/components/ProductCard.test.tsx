import { ProductCars } from "../../src/components/ProductCars";
import React from "react"
import renderer,{act} from "react-test-renderer"



describe("pruebas product cars", () => {


    test("prueba 1 snapshot de renderizado", () => {

        const wraper = renderer.create(
            <ProductCars product={{
                title: "cafe",
                id:1
            }}>
                {() => (
                    <h1>ProductCars</h1>
                )}
            </ProductCars>
        )

        expect(wraper.toJSON()).toMatchSnapshot()


    })

    test("prueba 1 snapshot de renderizado", () => {

        const wraper = renderer.create(
            <ProductCars product={{
                title: "cafe",
                id:1
            }}>
                {({count,increaseBy}) => (
                    <>
                    <h1>ProductCars</h1>
                    <span>{count}</span>
                    <button onClick={() => increaseBy(1)}>increase</button>
                    </>
                )}
            </ProductCars>
        )
        let three = wraper.toJSON()

        console.log((three as any).children)
        act(() => {
            (three as any).children[2].props.onClick()
        })

        three = wraper.toJSON()

        expect((three as any).children[1].children[0]).toBe("1")

    })




})