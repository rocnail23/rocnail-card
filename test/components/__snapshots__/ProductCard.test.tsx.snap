// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas product cars prueba 1 snapshot de renderizado 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    ProductCars
  </h1>
</div>
`;
